src/com/amazon/ion/impl/bin/IonRawBinaryWriter.java

```diff
@@ -543,23 +551,29 @@ private void updateLength(long length)
     private void pushContainer(final ContainerType type)
     {
         // XXX we push before writing the type of container
-        containers.push().initialize(type, buffer.position() + 1);
+        containers.push(c -> c.initialize(type, buffer.position() + 1));
     }
 
-    private void addPatchPoint(final long position, final int oldLength, final long value, final ContainerInfo container)
+    private void addPatchPoint(final ContainerInfo container, final long position, final int oldLength, final long value)
     {
+        // If we're adding a patch point we first need to ensure that all of our ancestors (containing values) already
+        // have a patch point. No container can be smaller than the contents, so all outer layers also require patches.
+        ListIterator<ContainerInfo> stackIterator = containers.iterator();
```

Comment:
This allocates a new `$Iterator` every time. It looks like we could store and reuse a single instance, which may improve performance.